### ISSUE TITLE
Real time autocomplete (emacs ido style or popup list)

### ISSUE THREAD

**Issue by purpleP** [fish-shell/fish-shell#3567]
I haven't used your shell, but AFAIK for some reason you still don't have this feature.

AFAIK you already have zsh-style tab completion that can show all available options right? So if I would type (| means cursor)
`~ cd ~|`
and hit `TAB` you will show all options like
`~ cd ~|
~/Downloads    ~/Documents
`
right?

So why not show them all always without me needing to press tab?
I'm using zsh now and it really bothers me that in the 21st century at post Intellij area I have to press tab to see options in my shell.
I mean, how is that even possible, are shell developers still leave in a cave or something?

On a serious note, this really hurts productivity if you're a fast touch typist.

So I'll explain what I'd like to see in shell.

```
~ tree
└── autocomplete
    ├── code
    │   ├── other
    │   └── some
    └── comics
        ├── bar
        └── foo

7 directories, 0 files

```
The shell always shows available alternatives inside `{ option1 | option2 | option3 }` and highlights the first option.
```
~ cd | { **autocomplete** } # ** means highlighting
```
pressing enter now means executing `cd autocomplete`, meaning enter would always choose the first option (thinking about it in that case its probably better to wait for at least one character typed).
`
~ cd a { **autocomplete** }
`
pressing `/`. For `cd` this means accept suggestion but the command isn't complete yet.
`
~ cd autocomplete/ { **code** | comics }
`
user can chose options with fuzzy-mathching.
`
~ cd autocomplete/cm {**comics**}
`
Again, pressing `/` accepts completion but not executes current command
`
~ cd autocomplete/comics/ {**bar**}
`
Pressing enter now executes `cd autocomplete/comics/bar` as desired.
With zsh I could also do this more effectively If I'm already familiar with content of the folders `cd ~/a/c/b TAB` would transform into `cd ~/autocomplete/comics/bar`, but that's very fragile thing and I rarely use it (doesn't work well with folder names containing `_-` or anything like that.
The number of keypresses it would take with my method is 8.
If I would do the same while having to press tab every time I want to see options or chose between them this would take 10 which isn't much of a difference in this case, but the difference gets bigger with a bigger number of options.

When the number of options is big almost every time I try to chose one by hitting tab multiple times I miss the options I need by one keypress and need to use arrow keys to get back and it's an awful experience.

One way in which this way of displaying options would fail is when the options list is very big and or when the options themselves are big like completions for command options. In this case it's better to use popup menu, but with same principles.

**Comment by krader1961** [comment]
Keep in mind that fish has two distinct mechanisms: auto-suggestions and completions. The former is based on you command history. The latter on a combination of auto-generated or hand crafted completion commands plus auto-generated pathnames. Personally I like the fish approach. I can't tell from the proposal exactly how you intend auto-suggestions and pathname completions to interact.

**Comment by purpleP** [comment]
@krader1961 Yeah, I know that. I use zsh-autosuggest plugin that emulates fish autosuggestions. And AFAIK autosuggestions aren't based only on history, but that doesn't matter. What matters is this.

1. Both history and context (context, not just pathname, pathname was just an example) autocompletion are useful.
2. User can be provided with choice of wether to use context, history or both and shell can show for example two separate popup or lists (ido style) with both type of completions. Notice that method that I've described doesn't require any additional keybindings, so user can use whaterver he chosen (like `C-f`) to accept history autosuggestion.

**Comment by krader1961** [comment]
> I haven't used your shell, but AFAIK for some reason you still don't have this feature.

If you haven't used fish why do you think you are in a position to propose a significant change in its behavior?

**Comment by ridiculousfish** [comment]
@purpleP Thank you for taking the time to write this up. I understand this as a request for automatic tab completions. Instead of having to press tab, the completions appear as a menu automatically, as you type.

The main reason fish doesn't do this is that tab completions may be very expensive. For example the `git` completion can invoke quite a lot of disk I/O. Autosuggestions are already viewed with suspicion, even though they work very hard to never create subprocesses, block the user's typing, etc. If fish were to invoke `git`, etc. as the user types, it would be very excessive.

However we can imagine showing a menu of the "cheap completions" (history, etc) as the user types - basically all the autosuggestions instead of the best one. This seems reasonable to me: web browsers already work this way, for example. We'd have to think hard about the relationship in the UI between these cheap completions and tab completions. The relationship is already confusing: I certainly hit tab when I want to hit right arrow. But maybe it can be done.

I'm skeptical that this can work but happy to be proven wrong, if someone were to prototype it. Or if zsh has this as a plugin I'd love to try it.

**Comment by purpleP** [comment]
@ridiculousfish Let's clarify something about expensive competitions. Disc IO on itself isn't a bad thing, right? So you could worry about three things:
1.Time it takes to get competitions.
2. Disk health.
3. Energy consumption.

The time doesn't matter, because the way this should work is when user typed character you would essentially create a Future that would have completions when it's done and if user typed a character before future is ready it's should be just cancelled. I mean if completions are already asynchronous then they wouldn't block ui, right?

Disk health for reads is probably an issue in HDD and it's already a previous generation tech. And Intel and Toshiba have next generation energy

**Comment by asukaminato0721** [comment]
Now here is a similar tool https://github.com/microsoft/inshellisense , but the completion is terrible. Since the completion is defined here https://github.com/withfig/autocomplete/blob/master/src/ls.ts , and missing a lot of things.

**Comment by arthsmn** [comment]
There's a zsh plugin for this: https://github.com/marlonrichert/zsh-autocomplete

**Comment by asukaminato0721** [comment]
now vscode support part of this by "terminal.integrated.suggest.enabled"

![Image](https://github.com/user-attachments/assets/588116ad-5117-433a-b5f6-b02c5925167c)

<https://x.com/code/status/1905380606864662913>

**Comment by alkene5** [comment]
https://github.com/user-attachments/assets/5f54aaa3-c688-41c9-a893-5fb29f1b5362

This feature really should be prioritized. The https://github.com/marlonrichert/zsh-autocomplete plugin has already implemented it on Zsh, and the effect in the video is exactly that.

**Comment by asukaminato0721** [comment]
I’ve impl this in  https://github.com/asukaminato0721/fish-shell/tree/3567 try it：）

```
set -g fish_autocomplete_autoshow 1
```

**Comment by mcauley-penney** [comment]
This is *the* most important feature of a shell for me. I was using zsh-autocomplete for a long time and I frankly think that every single shell should have this built-in, for all commands regardless of their expense. If there is room here for this to be implemented in core, set as opt-in, and gated behind an option, I'd like to involved in implementation. 

@asukaminato0721 Would you be interested in working together on this?

**Comment by asukaminato0721** [comment]
> This is _the_ most important feature of a shell for me. I was using zsh-autocomplete for a long time and I frankly think that every single shell should have this built-in, for all commands regardless of their expense. If there is room here for this to be implemented in core, set as opt-in, and gated behind an option, I'd like to involved in implementation.
> 
> [@asukaminato0721](https://github.com/asukaminato0721) Would you be interested in working together on this?

yeah, for me, this is a bit tricky to impl, feel free to ping me at issue there.

**Comment by mcauley-penney** [comment]
@asukaminato0721 Please try out [the changes I've made in my fork](https://github.com/asukaminato0721/fish-shell/compare/3567...mcauley-penney:fish-shell:3567) to see if this aligns with the behavior you expect. I think this is quite close to zsh-autosuggestions.

**Comment by GladioFeng** [comment]
@asukaminato0721 
I am very interested in this function and hope it will be released soon. 😊
